---
title: "Introduction to PhyloVision and Hotspot"
package: "`r BiocStyle::pkg_ver('BiocStyle')`"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Introduction to PhyloVision and Hotspot}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Preliminaries

If you have yet to install VISION, we recommend installing the package from Github to install this package. Full source code can be found at the VISION Github repository, available [here](http://www.github.com/YosefLab/VISION).

```r
require(devtools)
install_github("YosefLab/VISION")
```

Once VISION and R are installed, you may load in VISION using `library(VISION)`.

# Using PhyloVision

Running an analysis with vision consists of three steps:

1. Creating the PhyloVision object
2. Running the `phyloAnalyze` function
3. Running Hotspot
4. Browsing results

First, we need to load VISION, reticulate and ape.
```{r setup, eval=F}
knitr::opts_chunk$set(echo = TRUE)
library(VISION)
library(reticulate)
library(ape)
```

## Creating a PhyloVision Object with a Tree

Vision objects now support dendrograms for visualization and analysis. To create the PhyloVision object, you need all of the same data, an expression matrix, and metadata and/or signatures, but you also include a tree object from the Ape package.

```{r create, eval=F}
# Read the expression and meta data
file_path <- "data/embryogenesis" # Replace the path

expr <- read.table(paste(file_path, "expr_filtered.tsv.gz", sep="/"), check.names = FALSE, sep = "\t")

meta <- read.table(paste(file_path, "meta.tsv", sep="/"), check.names = FALSE, sep = "\t", row.names = 1, header=TRUE)
meta$Cell_Type <- as.factor(meta$Cell_Type)
# Signature file
sig <- paste("data", "h.all.v5.2.symbols.gmt", sep="/")

# Read the tree
tree <- read.tree(paste(file_path, "embryo_tree.newick", sep="/"))
# Collapse one mutations
tree <- collapse.singles(tree)
expr <- expr[, tree$tip.label]

# Construct the Vision object
vis <- PhyloVision(tree=tree, data=expr, signatures=sig, meta=meta, num_neighbors=30, projection_genes= rownames(expr))
```
**Tree**

The `ape` phylogeny object. Singleton edges should be collapsed prior to use. The tree should be filtered to have the same cell (leaf) names as the expression data.

**Expression Data**

The provided expression data should be scaled and normalized. It is recommended to apply more advanced normalization procedures such as batch correction or removal of technical confounders.

The expression data should not be log-transformed prior to loading into VISION.

**Signatures**

Signatures can be provided as a list of paths to signature files (\*.gmt) or Signature objects.

See [the signature vignette](Signatures.html) for more information on finding or creating gene Signatures.

**Meta Data**

An R data.frame with cell-level meta-data.  This could be confounding variables (e.g. percentage of mitochondrial RNA, number of genes detected) or experimental covariates (e.g. genotype, donor, batch).

This input is optional if Signatures are provided.

**Other Options**

Other options and inputs can be provided to customize how VISION runs.

## Running PhyloVision analysis

Next, we can perform the normal Vision analysis using the tree as the latent space.

``` {r analyze, eval=F}
vis <- phyloAnalyze(vis)
```
## Running Hotpost analysis

We can also perform Hotspot module analysis. The expression data is already logged for us.
```{r hotspot, eval=F}
vis <- runHotspot(vis, model="normal", tree=TRUE, min_gene_threshold=50, n_neighbors = 30, number_top_genes = nrow(expr), logdata=FALSE)
```

The full Hotspot API is exposed for analysis as well. For more on the Hotspot API see [here](https://yoseflab.github.io/Hotspot/index.html)
```{r full_h, eval=FALSE}
# Init Hotspot
hs <- hsInit(vis, model = "normal", tree=TRUE, logdata=FALSE)
# Init Hotspot KNN
hs <- hsCreateKnnGraph(hs, vis, n_neighbors=30)
# perform Hotspot analysis and store results in R
hs_genes <- hsComputeAutoCorrelations(hs, number_top_genes=12438)
# Compute localcorr
hs <- hsComputeLocalCorrelations(hs, hs_genes)
# Calculate Hotspot Module Scores for informative genes
hs <- hsCalculateModuleScores(hs)
# Cluster Hotspot modules and perform Vision based analysis on HS Modules and 
vis <- analyzeHotspotObjectVision(vis, hs, tree=TRUE)
```

We can also access further Hotspot functionality using Reticulate and Python.
```{r inspectModules, eval=F}
hs <- loadHotspotObject(bytes=vis@Hotspot)
library(reticulate)
use_python('/usr/bin/python3')
```
```{python modulesPlot, eval=F}
import matplotlib.pyplot as plt
import hotspot
hs.plot_local_correlations()
plt.show()
```
Note: the heatmap can also be visualized like so:
```{r heatmapR, eval=F}
hs <- loadHotspotObject(bytes=vis@Hotspot)
library(paletteer)
draw_hotspot_heatmap(hs)
```

## Viewing Results

Finally, we can launch the Vision browser.
```{r eval=FALSE}
viewResults(vis)
```


